elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -11,21 +11,63 @@
  *
  * Removal or modification of this copyright notice is prohibited.
  */
+/* eslint-disable no-bitwise */
+/* eslint-disable @typescript-eslint/prefer-for-of */
 
+import { hash } from '@liskhq/lisk-cryptography';
 import { Proof } from './types';
+import { EMPTY_HASH, LEAF_PREFIX, BRANCH_PREFIX } from './constants';
 
 export class MerkleTree {
-	// eslint-disable-next-line
 	private _root: Buffer;
-	// eslint-disable-next-line
-	public constructor(_initValues: Buffer[]) {
-		this._root = Buffer.alloc(0);
+	private _dataLength = 0;
+	private _data: { [key: string]: Buffer } = {};
+
+	public constructor(initValues: Buffer[] = []) {
+		this._dataLength = initValues.length;
+		if (this._dataLength === 0) {
+			this._root = EMPTY_HASH;
+			return;
+		}
+		const leafHashes = new Array<Buffer>(this._dataLength);
+		for (let i = 0; i < this._dataLength; i += 1) {
+			const value = Buffer.concat(
+				[LEAF_PREFIX, initValues[i]],
+				LEAF_PREFIX.length + initValues[i].length,
+			);
```

Comment:
@shuse2 Does not it be: 

```
const value = Buffer.concat(
   LEAF_PREFIX
   initValues[i],
   LEAF_PREFIX.length + initValues[i].length,
);
```

Reply:
Buffer.concat is `concat(value: Buffer[], totalLength: number)`
so the above is correct, and total length is just there to make it efficient

Reply:
Ok got it. But I don't think providing length which is an optional parameter makes it efficient. It's just there if one wants to truncate the result buffer for a particular length.

Reply:
> If totalLength is not provided, it is calculated from the Buffer instances in list by adding their lengths.
It's there not to recalculate the total size of buffer while concatenating, so the performance will be better if you know the size ahead